### PR TITLE
feat: in-memory canvas snapshots

### PR DESCRIPTION
**Pull request by RobertasJ**

### CODE REVIEW

**Review by marc2332 — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 1]
```diff
@@ -3,4 +3,5 @@ Cargo.lock
 .idea
 snapshot_before.png
 snapshot_after.png
-documents_example
\ No newline at end of file
+documents_example
+bacon.toml
```
Comment (marc2332) [inline .gitignore:7]: wot?

Comment (marc2332) [reply .gitignore:7]: https://dystroy.org/bacon/ ?

Comment (RobertasJ) [reply .gitignore:7]: yep its that, rust-analyzer doesnt seem to work well

File: crates/testing/src/test_handler.rs [hunk 2]
```diff
@@ -300,7 +300,7 @@ impl TestingHandler {
     }
 
     /// Render the app into a canvas and save it into a file.
-    pub fn save_snapshot(&mut self, snapshot_path: impl Into<PathBuf>) {
+    pub fn create_snapshot(&mut self) -> Vec<u8> {
```
[the remark this answers is not in the document]
Comment (marc2332) [reply crates/testing/src/test_handler.rs:303]: ah hold, we probably can't do this, feel free to return SkData instead

File: crates/testing/src/test_handler.rs [hunk 3]
```diff
@@ -345,12 +345,16 @@ impl TestingHandler {
             .encode(context.as_mut(), EncodedImageFormat::PNG, None)
             .expect("Failed to encode the snapshot.");
 
-        // Save snapshot
+        snapshot_data.as_bytes().to_vec()
```
[the remark this answers is not in the document]
Comment (RobertasJ) [reply crates/testing/src/test_handler.rs:348]: this returns a dangling refference

[the remark this answers is not in the document]
Comment (marc2332) [reply crates/testing/src/test_handler.rs:348]: > this returns a dangling refference

https://github.com/marc2332/freya/pull/979#discussion_r1798147113

File: crates/testing/src/test_handler.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (RobertasJ) [reply crates/testing/src/test_handler.rs:303]: I cant access the type to put it into the function definition

[the remark this answers is not in the document]
Comment (marc2332) [reply crates/testing/src/test_handler.rs:303]: You might need to reexport it and mock it, check skia.rs and mocked.rs

[the remark this answers is not in the document]
Comment (RobertasJ) [reply crates/testing/src/test_handler.rs:303]: mocked.rs has a mocked Data type, but im not very familiar with mocking so i am not sure if i should use it

[the remark this answers is not in the document]
Comment (marc2332) [reply crates/testing/src/test_handler.rs:303]: Hold on, SkData is already exported, its named `Data`, no need to modify mocked.rs or skia.rs

File: crates/testing/src/test_handler.rs [hunk 4]
```diff
@@ -341,16 +342,18 @@ impl TestingHandler {
...
-            .expect("Failed to encode the snapshot.");
+            .expect("Failed to encode the snapshot.")
+    }
 
-        // Save snapshot
+    pub fn save_snapshot(&mut self, snapshot_path: impl Into<PathBuf>) {
         let mut snapshot_file =
             File::create(snapshot_path.into()).expect("Failed to create the snapshot file.");
-        let snapshot_bytes = snapshot_data.as_bytes();
+        let data = self.create_snapshot();
```
Comment (marc2332) [inline crates/testing/src/test_handler.rs:353]:
```suggestion
        let snapshot_data = self.create_snapshot();
```

File: crates/testing/src/test_handler.rs [hunk 5]
```diff
@@ -341,16 +342,18 @@ impl TestingHandler {
...
+
         snapshot_file
-            .write_all(snapshot_bytes)
+            .write_all(&data)
```
Comment (marc2332) [inline crates/testing/src/test_handler.rs:356]:
```suggestion
            .write_all(&snapshot_data )
```

**Review by marc2332 — CHANGES_REQUESTED** [review]

**Review by marc2332 — APPROVED** [review]

**Pull request merged**